<a href="https://colab.research.google.com/github/jsmazorra/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Johan_Mazorra_LS_DS13_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [6]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-02-17 21:38:24--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.1’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-02-17 21:38:24 (285 KB/s) - ‘house-votes-84.data.1’ saved [18171/18171]



In [7]:
#After loading the data, we're going to import pandas and organize the columns.
import pandas as pd

vd = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

vd.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [13]:
# Since in the preview above we saw that the data displays the 'NaNs' as '?', then we're going to use the code below 
# to transform them into NaNs so it would be easier to clean up.
vd = vd.replace({"y": 1, "n": 0, "?": np.NaN})

vd.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [14]:
# Now we could see all the missing values from the dataframe.

vd.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [15]:
# Now I'm going to make another df for the republican votes and the democrat votes so we can use them for this exercise.

rep = vd[vd['party']=='republican']

rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [16]:
dem = vd[vd['party']=='democrat']

dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [19]:
vd['party'.value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [98]:
# Everything seems to be in place, now I just want to check the stats of both the republican and democrat votes.

print(rep.shape)
rep.describe()

(168, 17)


handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [74]:
print(dem.shape)
dem.describe()

(267, 17)


handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

# ***On Republican and Democratic support of the budget bill***

---



---



Null Hypothesis: Republicans support the budget bill as much as Democrats.

Alternative Hypothesis: Republican support for this bill is not the same as Democrats.

Confidence Level: 99%, my critical value is of .01

In [96]:
from scipy import stats

stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

# **Conclusion**

---

T Statistic: -23.2; P-value: ~0.0000000021

Based on a t-stat of -23.2 and a p-value of 0, I reject the null hypothesis where Republicans support the budget bill as much as the Democrats; so I suggest the alternative hypothesis.


# ***On Republican and Democratic support of the immigration bill***

---

---



Null Hypothesis: Republican support for the immigration bill will be as much as the Democrats.

Alternative Hypothesis: Republican support for this bill is not as much as the Democrats.

Confidence Level: 90%, my critical value is of .1

In [101]:
stats.ttest_ind(dem['immigration'], rep['immigration'], nan_policy='omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

# ***Conclusion***

---

T Statistic: -1.74; P-value: 0.08

Based on the t-stat of -1.74 and p-value of 0.08, it is safe to conclude that I fail to reject the null hypothesis where Republicans support the immigration bill as much as Democrats.

In [43]:
dem['immigration'].mean()

0.4714828897338403

# ***On Republican and Democratic support of the physician fee freeze bill***

---

---



Null Hypothesis: Republican support for the physician fee freeze bill will be as much as the Democrats.

Alternative Hypothesis: Republican support for this bill is not as much as the Democrats.

Confidence Level: 99%, my critical value is of .01

In [103]:
stats.ttest_ind(dem['physician-fee-freeze'], rep['physician-fee-freeze'], nan_policy='omit')

Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

# ***Conclusion***

---



T-Statistic: -49.4; P-value: ~0

Based on the t-stat of -49.4 and p-value of 0, I reject the null hypothesis where Republicans support the physician fee freeze bill as much as the Democrats. So I suggest the alternative hypothesis.

# **Stretch Goals**

Refactor your code into functions so it's easy to rerun with arbitrary variables

In [147]:
# For the stretch goal, I'm going to refactor my code into functions.

from scipy.stats import ttest_1samp, ttest_ind
def one_sample_ttest(array):
  p_value = ttest_1samp(array, popmean=0.47, nan_policy='omit')[1]
  if p_value < 0.05:
    print('The p_value of the given array compared to the expected mean is less than the critical value of 0.05 therefore, we will reject our Null Hypothesis and suggest the Alternative Hypothesis')
  else:
    print('The p_value of the given array compared to the expected mean is greater than or equal to the critical value of 0.05 therefore, we cannot reject the Null Hypothesis')
    
# Let's see if it's working.
print(one_sample_ttest(stats.ttest_1samp(dem['immigration'], 0.47, nan_policy='omit')))

The p_value of the given array compared to the expected mean is greater than or equal to the critical value of 0.05 therefore, we cannot reject the Null Hypothesis
None


In [145]:
# And just to double check that it's functioning properly, I'm going to get the t-test result.
# Everything looks to be in order.

stats.ttest_1samp(dem['immigration'], 0.47, nan_policy='omit')

Ttest_1sampResult(statistic=0.04808360379881848, pvalue=0.9616862459950337)